Remove unprotect calls that rchk claims to be over unprotecting.

astamm opened the pull request:
While checking against `rhub/rchk` the {phutil} package I am about to submit to CRAN, it spotted:

```bash
Function cpp11::detail::store::insert(SEXPREC*)
  [PB] has negative depth /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/protect.hpp:302
  [UP] attempt to unprotect more items (2) than protected (0), results will be incomplete /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/protect.hpp:302
  [PB] has possible protection stack imbalance /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/protect.hpp:305

Function cpp11::writable::r_vector<double>::reserve_data(SEXPREC*, bool, long)
  [PB] has negative depth /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1342
  [UP] attempt to unprotect more items (2) than protected (0), results will be incomplete /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1342
  [PB] has possible protection stack imbalance /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1343

Function cpp11::writable::r_vector<double>::resize_data(SEXPREC*, bool, long)
  [PB] has negative depth /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1367
  [UP] attempt to unprotect more items (1) than protected (0), results will be incomplete /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1367
  [PB] has possible protection stack imbalance /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1368

Function cpp11::writable::r_vector<double>::resize_names(SEXPREC*, long)
  [PB] has negative depth /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1389
  [UP] attempt to unprotect more items (1) than protected (0), results will be incomplete /github/home/R/x86_64-pc-linux-gnu-library/4.6/cpp11/include/cpp11/r_vector.hpp:1389
```

Full details here: https://github.com/tdaverse/phutil/actions/runs/14893316603/job/41830462913.

I therefore temporarily vendored {cpp11} headers and removed the `UNPROTECT()` calls and now `rchk` does not complain anymore.

This PR proposes their removal but maybe it needs more tests/investigation.

astamm commented:
Though it seems to break the CI because `FAQ` vignette fails with these modifications.
Maybe it was a false positive from `rchk`.
If you can shed some light on this, I would very much appreciate it. Thanks.

DavisVaughan commented:
These unprotects look correct to me. You can see the matching `PROTECT()` above them.

astamm commented:
Yes, but still `rchk` complains.

I found a fix for `insert` in `protect.hpp`.

The key changes were to remove PROTECT/UNPROTECT calls since:

- `list` from `get()` is already preserved;
- The cons cell is part of the preserve list structure;
- The input `x` is protected by the caller.

I have not yet managed to find the issue in `r_vector.hpp`.

astamm commented:
I found the fix for `r_vector.hpp`. In summary,

#### Key changes to `reserve_data()`

1. Removed all explicit PROTECT/UNPROTECT calls since protection is handled by the `safe` wrapper in both `resize_data` and `resize_names()`
2. Simplified the flow of the names handling
3. Let R's garbage collection handle the intermediate objects

#### Key change to `resize_data()`

1. Moving both the PROTECT and UNPROTECT calls since the `safe` wrapper around `Rf_allocVector` already handles protection.

#### Key changes to `resize_names()`

1. The `safe` wrapper around `Rf_allocVector` already handles protection
2. Removed direct pointer access via STRING_PTR_RO
3. Use STRING_ELT instead of direct pointer access

DavisVaughan commented:
The `safe` wrapper does not handle protection when the output type is `SEXP`. The caller must still protect the `SEXP` output. There must be a problem with rchk, possibly only with r-devel, the existing code looks fine to me.
